hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -0,0 +1,380 @@
+import com.hedera.node.config.data.LedgerConfig;
+import com.hedera.node.config.data.TokensConfig;
+import com.swirlds.base.utility.Pair;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link HederaFunctionality#TOKEN_REJECT}.
+ * This transaction type is used to reject tokens from an account and send them back to the treasury.
+ */
+@Singleton
+public class TokenRejectHandler extends BaseTokenHandler implements TransactionHandler {
+
+    @Inject
+    public TokenRejectHandler() {}
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenRejectOrThrow();
+
+        // If owner account is specified, verify that it's valid and require that it signs the transaction.
+        // If not specified, payer account is considered as the one rejecting the tokens.
+        if (op.hasOwner()) {
+            final var accountStore = context.createStore(ReadableAccountStore.class);
+            verifyOwnerAndRequireKey(op.ownerOrThrow(), context, accountStore);
+        }
+    }
+
+    /**
+     * Verifies that the owner account exists and ensures the account's key is valid and required for the transaction.
+     *
+     * @param ownerId The AccountID of the owner whose key needs to be validated.
+     * @param context The PreHandleContext providing transaction context.
+     * @param accountStore The store to access readable account information.
+     * @throws PreCheckException If the sender's account is immutable or the sender's account ID is invalid.
+     */
+    private void verifyOwnerAndRequireKey(
+            @NonNull final AccountID ownerId,
+            @NonNull final PreHandleContext context,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+
+        final var ownerAccount = accountStore.getAliasedAccountById(ownerId);
+        validateTruePreCheck(ownerAccount != null, INVALID_OWNER_ID);
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = ownerAccount.key();
+        if (key == null || !isValid(key)) {
+            throw new PreCheckException(ACCOUNT_IS_IMMUTABLE);
+        }
+        context.requireKey(key);
+    }
+
+    @SuppressWarnings("java:S2259")
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn, "Transaction body cannot be null");
+        final var op = txn.tokenRejectOrThrow();
+
+        validateFalsePreCheck(op.rejections().isEmpty(), INVALID_TRANSACTION_BODY);
```

Comment:
We could probably add `EMPTY_TOKEN_REJECTIONS` response code

Reply:
Okay. Will add it to the protobufs and include it :)

Reply:
Done, I've added it. :)